### DISCUSSION TITLE
Fabric-history undo erasing

### DISCUSSION THREAD

**Discussion by hanadinh** [fabricjs/fabric.js#8851] [Q&A] [answered]
I'm implementing a freehand drawing board using EraserBrush for eraser and [fabric-history](https://github.com/alimozdemir/fabric-history) library. However, I found out that when you erase parts of the objects and undo it, the drawing doesn't go back to the previous state as expected. It just goes back to the state before you erase (acting like erasing action never happens). 

Can someone let me know how to fix this issue so the 2 libraries can work together? 

I also saw this [discussion](https://github.com/fabricjs/fabric.js/discussions/7582) to implement undo/redo ourselves. However I don't really understand the answer. Can someone give me an example of the implementation?

**Comment by ShaMan123** [comment, marked as answer]
I don't know fabric-history but I don't trust a generalized solution for history in fabric (we, the team, discussed and couldn't think of a good enough solution, it varies a lot)
You should hook to events and handle it.

**Reply by ShaMan123** [reply]
More elaborately, you can removed add paths from the eraser of an object (don't forget to mark it dirty and to render)
